test/shared/src/main/scala/zio/test/Spec.scala

```diff
@@ -347,11 +347,21 @@ final case class Spec[-R, +E](caseValue: SpecCase[R, E, Spec[R, E]]) extends Spe
             .flatMap(r => scoped.map(_.provideEnvironment(r)).provideSomeEnvironment[Scope](r.union[Scope]))
         )
       case MultipleCase(specs) =>
-        Spec.scoped[R0](
-          layer.memoize.flatMap(layer =>
-            layer.mapError(TestFailure.fail).build.map(_ => Spec.multiple(specs.map(_.provideLayer(layer))))
-          )
-        )
+        Spec.scoped[R0] {
+          ZIO.scopeWith { scope =>
+            import Spec.RichLayer
+
+            val scopedLayer: ZLayer[R0, E1, R] =
+              ZLayer.makeSome[R0, R](
+                ZLayer.succeed(scope),
+                layer.extendScope
+              )
+
+            scopedLayer.memoizeWithinScope(scope).map { memoizedLayer =>
+              Spec.multiple(specs.map(_.provideLayer(memoizedLayer)))
+            }
+          }
+        }
```

Comment:
Let's try to reuse `memoize` to avoid code duplication. Also let's chain the methods together, creating a separate val for the scoped layer feels like it should be used more than once:

```scala
        Spec.scoped[R0] {
          ZIO.scopeWith { scope =>
            ZLayer
              .makeSome[R0, R](
                ZLayer.succeed(scope),
                layer.extendScope
              )
              .memoize
              .map { memoizedLayer =>
                Spec.multiple(specs.map(_.provideLayer(memoizedLayer)))
              }
          }
        }
```